server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -470,6 +491,50 @@ private static boolean upgradeLegacyNodeFolders(Logger logger, Settings settings
         return true;
     }
 
+    /**
+     * Checks to see if we can upgrade to this version based on the existing index state. Upgrading
+     * from older versions can cause irreversible changes if allowed.
+     * @param settings
+     * @param xContentRegistry
+     * @param logger
+     * @param nodePaths
+     * @throws IOException
+     */
+    private static void checkForIndexCompatibility(
+        Settings settings,
+        NamedXContentRegistry xContentRegistry,
+        Logger logger,
+        NodePath... nodePaths
+    ) throws IOException {
+        final Path[] paths = Arrays.stream(nodePaths).map(np -> np.path).toArray(Path[]::new);
+        NodeMetadata metadata = PersistedClusterStateService.nodeMetadata(paths);
+
+        PersistedClusterStateService persistedClusterStateService = new PersistedClusterStateService(
+            paths,
+            metadata.nodeId(),
+            xContentRegistry,
+            new ClusterSettings(settings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS),
+            () -> TimeValue.nsecToMSec(System.nanoTime())
+        );
+
+        PersistedClusterStateService.OnDiskState onDiskState = persistedClusterStateService.loadBestOnDiskState();
```

Comment:
I think this is too early in the node's lifecycle for loading the cluster state (that's why you had to create your own `NamedXContentRegistry`). Instead we should record the oldest index version in `NodeMetadata`, which means adding an entry to the `userData` attached to the Lucene commit for the cluster state.